tools/cmake/mbed_greentea.cmake

```diff
@@ -22,39 +22,39 @@ include(${CMAKE_CURRENT_LIST_DIR}/app.cmake)
 
 macro(mbed_greentea_add_test)
     set(options)
-    set(singleValueArgs TEST_NAME)
+    set(singleValueArgs 
+        TEST_NAME
+        TEST_MAIN)
     set(multipleValueArgs
         TEST_INCLUDE_DIRS
         TEST_SOURCES
-        TEST_REQUIRED_LIBS
-    )
+        TEST_REQUIRED_LIBS)
     cmake_parse_arguments(MBED_GREENTEA
         "${options}"
         "${singleValueArgs}"
         "${multipleValueArgs}"
         ${ARGN}
     )
 
-    set(TEST_NAME ${MBED_GREENTEA_TEST_NAME})
-
     add_subdirectory(${MBED_PATH} build)
 
-    add_executable(${TEST_NAME})
-
-    # Explicitly enable BUILD_TESTING until CTest is added to the Greentea client
+    add_executable(${MBED_GREENTEA_TEST_NAME}) # Explicitly enable BUILD_TESTING until CTest is added to the Greentea client
     set(BUILD_TESTING ON)
 
-    mbed_configure_app_target(${TEST_NAME})
+    mbed_configure_app_target(${MBED_GREENTEA_TEST_NAME})
 
-    target_include_directories(${TEST_NAME}
+    target_include_directories(${MBED_GREENTEA_TEST_NAME}
         PRIVATE
             .
             ${MBED_GREENTEA_TEST_INCLUDE_DIRS}
     )
 
-    target_sources(${TEST_NAME}
+    if("${MBED_GREENTEA_TEST_MAIN}" STREQUAL "")
+        set(MBED_GREENTEA_TEST_MAIN main.cpp)
+    endif()
+    target_sources(${MBED_GREENTEA_TEST_NAME}
         PRIVATE
-            main.cpp
+            ${MBED_GREENTEA_TEST_MAIN} 
             ${MBED_GREENTEA_TEST_SOURCES}
     )
```

Comment:
For other reviewers' reference only:

The old assumption of `main.cpp` is not always true - we shouldn't hardcode any file names here or prevent a test from using C or having a different name. In my opinion, two better designs could be:
* Option 1: Default to `main.cpp` only when the optional `TEST_SOURCES` is unset. If `TEST_SOURCES` is set by a test, the test should include the main program in this parameter.
* Option 2: Don't assume `main.cpp`, and make `TEST_SOURCES` compulsory.

Either option would require changing CMake definitions of existing tests which are many. So to avoid doing that now, a new parameter `TEST_MAIN` is added in this PR, though obviously having both `TEST_MAIN` and `TEST_SOURCES` is largely redundant. Maybe this can be refactored in the future.

Reply:
Making TEST_SOURCES compulsory sounds like a good approach to me. If we want to do this, we should refactor now as part of this work. I don't think it's an important enough of a change we'd get to it later.

Reply:
> Making TEST_SOURCES compulsory sounds like a good approach to me. If we want to do this, we should refactor now as part of this work. I don't think it's an important enough of a change we'd get to it later.

I can update my PR with this refactoring, if that's what makes sense to everyone else

Reply:
Given that there're a very large number of tests, maybe we can create a separate PR to do that, and have this PR depend on that?

The change itself should be straightforward, i.e. use
```
grep -rnI mbed_greentea_add_test
```
to find all occurrences.

Reply:
@LDong-Arm @Patater I have made TEST_SOURCES compulsory and updated all greentea tests accordingly. The changes are rebased into this PR.